### PR TITLE
DM-31888: swap server and client: client in the CSC, server in the mock controller

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

File: python/lsst/ts/hexrotcomm/command_telemetry_client.py [hunk 1]
```diff
@@ -65,46 +75,53 @@ class CommandTelemetryClient:
...
         self.telemetry_writer = None
-        self.command_loop_task = asyncio.Future()
-        self.command_loop_task.set_result(None)
-        self.telemetry_loop_task = asyncio.Future()
-        self.telemetry_loop_task.set_result(None)
         self.connect_task = asyncio.create_task(self.connect())
+        # The state of (command_connected, telemetry_connected)
+        # the last time connect_callback was called.
+        self.was_connected = (False, False)
+        self.should_be_connected = False
```
Comment (teweitsai) [inline python/lsst/ts/hexrotcomm/command_telemetry_client.py:111]: Did you add the doc string for these two attributes in `Parameters`?

File: python/lsst/ts/hexrotcomm/command_telemetry_client.py [hunk 2]
```diff
@@ -65,46 +75,53 @@ class CommandTelemetryClient:
...
+        # by the data type of Command.counter.
+        self._command_counts = dict()
+
+        # Task the user can set to an asyncio.Future()
+        # in order to detect when the next telemetry is read.
+        # Only intended for unit tests.
+        self._telemetry_task = asyncio.Future()
+
+        self.read_telemetry_and_config_task = utils.make_done_future()
+        self.monitor_command_reader_task = utils.make_done_future()
```
Comment (teweitsai) [inline python/lsst/ts/hexrotcomm/command_telemetry_client.py:124]: Did you add the doc string for these two attributes in `Parameters`? Actually, I felt there are too many attributes in this class right now...

File: python/lsst/ts/hexrotcomm/command_telemetry_client.py [hunk 3]
```diff
@@ -134,182 +151,204 @@ async def close(self):
...
-            Command to run.
+            Command to write. Its counter field will be set.
         """
-        raise NotImplementedError()
+        if not self.command_connected:
+            raise RuntimeError("Command socket not connected")
+        if not isinstance(command, structs.Command):
+            raise ValueError(
+                f"command={command!r} must be an instance of structs.Command"
+            )
```
Comment (teweitsai) [inline python/lsst/ts/hexrotcomm/command_telemetry_client.py:347]: Maybe add the doc string of `RuntimeError` and `ValueError`?

File: python/lsst/ts/hexrotcomm/base_csc.py [hunk 4]
```diff
@@ -27,34 +27,66 @@
...
+                state_transition_dict[(beg_state, end_state)] = []
+                continue
+            step = 1 if end_ind > beg_ind else -1
+            command_state_list = []
+            for next_ind in range(beg_ind, end_ind, step):
+                from_state = ordered_states[next_ind]
+                to_state = ordered_states[next_ind + step]
+                command = basic_state_transition_commands[from_state, to_state]
+                command_state_list.append((command, to_state))
+            state_transition_dict[(beg_state, end_state)] = command_state_list
```
Comment (teweitsai) [inline python/lsst/ts/hexrotcomm/base_csc.py:81]: Please consider to decrease the indentation of `for` loop here.

File: python/lsst/ts/hexrotcomm/base_csc.py [hunk 5]
```diff
@@ -270,16 +269,18 @@ def assert_commandable(self):
...
-            raise salobj.ExpectedError("Controller is not connected")
+        self.assert_connected()
         if not self.evt_commandableByDDS.data.state:
             raise salobj.ExpectedError(
                 "Controller has CSC commands disabled; "
                 "use the EUI to enable CSC commands"
             )
 
+    def assert_connected(self):
+        """Assert that the CSC is connected to the low-level controller."""
```
Comment (teweitsai) [inline python/lsst/ts/hexrotcomm/base_csc.py:280]: Maybe add the doc string of `ExpectedError`?

**Review by teweitsai — APPROVED** [review]
The update looks good to me. Thanks for this change!

File: doc/version_history.rst [hunk 6]
```diff
@@ -6,6 +6,41 @@
...
+    * Connect to the low-level controller as part of the ``start`` command.
+    * Make the CSC summary state mostly independent of the low-level controller state (an excellent suggestion from Tiago).
+      As part of the ``enable`` command, the CSC commands the low-level controller to its own enabled state,
+      including trying to clear errors, if necessary.
+      While the CSC is enabled:
+
+        * If the low-level controller goes to fault state then the CSC also goes to fault state.
+          It remains connected, so you can continue to see what the low-level controller is reporting.
+        * If the low-level controller leaves enabled state or the EUI takes control, the CSC goes to disabled state.
+        * If the CSC loses connection to the low-level controller, the CSC goes to fault state.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/version_history.rst:27]: Good point. I overhauled the "communication protocol" documentation (which I had forgotten to update!) and replaced some of the details from the version history with a link.

File: python/lsst/ts/hexrotcomm/command_telemetry_client.py (hunk 1 above)
Comment (r-owen) [reply python/lsst/ts/hexrotcomm/command_telemetry_client.py:111]: Good point (though they are not constructor parameters). I will add an Attributes section and include these (as well as all the other public attributes).

File: python/lsst/ts/hexrotcomm/command_telemetry_client.py (hunk 2 above)
Comment (r-owen) [reply python/lsst/ts/hexrotcomm/command_telemetry_client.py]: Yes. I agree. I made some private and documented the rest in a new Attributes section.

**Comment by r-owen** [comment]
I also added an Attributes section to the CommandTelemetryServer class (which fortunately, has a more reasonable number of public attributes than CommandTelemetryClient).

**Pull request merged**